horovod/common/ops/adasum_cuda_ring.h

```diff
@@ -0,0 +1,113 @@
+// Copyright 2019 Microsoft. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#ifndef HOROVOD_ADASUM_CUDA_RING_H
+#define HOROVOD_ADASUM_CUDA_RING_H
+
+#include "mpi.h"
+
+#include "../common.h"
+#include "../global_state.h"
+#include "../mpi/mpi_context.h"
+#include "gpu_operations.h"
+#include "cuda_fp16.h"
+#include "adasum_cuda_kernels.h"
+
+namespace horovod {
+namespace common {
+
+struct Ring {
```

Comment:
How does the perf compare if we put the give a stream to each ring? Since rings are reducing different tensors, giving them separate rings might make more sense.

Reply:
I tried using multiple streams previously, but based on layerids instead of rings. Since the kernel computation is very cheap and most of the overhead comes from the dispatching host thread being blocked by the launch of the new kernel if it's not sent to a separate stream, I think using one stream is fine. But I will try to quickly experiment with using different streams for different rings just to check.

Reply:
Ok I checked this, and using a different stream for every ring makes it worse. So I went back to using a single stream for all of them, but having it be stored in the allrings.